tiledb/common/evaluator/evaluator.h

```diff
@@ -0,0 +1,411 @@
+    }
+  };
+
+  struct OndemandValue {
+    std::shared_future<return_type> value;
+    uint32_t num_readers;
+    OndemandValue() = delete;
+    OndemandValue(const std::shared_future<return_type>& v)
+        : value(v)
+        , num_readers(0) {
+    }
+  };
+
+  /**
+   * Enforces the cache policy by evicting entries if necessary.
+   * This operation assumes the bookkeeping mutex is already locked.
+   */
+  virtual void enforce_policy(const Value& v) = 0;
+
+  /**
+   * Double linked list where the head is the least
+   * recently used element from the cache.
+   */
+  List lru_;
+
+  /**
+   * Each entry stores a promised value which is in the process of
+   * being fetched. Readers will wait on it to become available.
+   */
+  std::unordered_map<Key, OndemandValue> inprogress_;
+
+  /**
+   * Map of cached entries by key. Each entry stores and iterator
+   * within the LRU list and a shared_ptr to the value.
+   * If an entry is in this map, it is guaranteed that the value is
+   * already available to serve readers.
+   */
+  std::unordered_map<Key, Cached_Entry> entries_;
+
+  /**
+   * Bookkeeping lock. This covers the atomocity of operations over the three
+   * containers above.
+   */
+  std::mutex mutex_;
+};
+
+/**
+ * Passing this policy to the evaluator will configure it to
+ * always evaluate the callback function and never cache the results.
+ *
+ * @tparam Key The type of key
+ * @tparam Value The type of value
+ */
+template <class Key, class Value>
+class ImmediateEvaluation {
+ public:
+  using key_type = Key;
+  using value_type = Value;
+  using return_type = std::shared_ptr<Value>;
+
+  /**
+   * Evaluates the function and constructs a shared_ptr with the prduced value.
+   *
+   * @tparam F The callback function type.
+   * @param f The callback function to evaluate.
+   * @param key The key to evaluate the callback on.
+   */
+  template <class F>
+  return_type operator()(F f, const Key& key) {
+    return make_shared<value_type>(f(key));
+  }
+};
+
+/**
+ * Policy that enforces a maximum number of entries in the cache.
+ * Once the threshold is reached, the LRY entry is evicted making space
+ * for a new value to evaluate.
+ *
+ * This policy is useful for testing purposes.
+ *
+ * @tparam Key The type of key
+ * @tparam Value The type of value
+ */
+template <class Key, class Value, size_t N>
+class MaxEntriesCache : public virtual CachePolicyBase<Key, Value> {
+ public:
+  using return_type = std::shared_ptr<Value>;
+
+  MaxEntriesCache()
+      : num_entries_(0) {
+    if constexpr (N == 0) {
+      throw std::logic_error(
+          "The maximum number of entries must be greater than zero.");
+    }
+  }
+
+  virtual void enforce_policy(const Value&) override {
+    if (num_entries_ == max_entries_) {
+      auto evicted_value = this->evict_lru();
+      --num_entries_;
+      return;
+    }
+    ++num_entries_;
+  }
+
+ private:
+  size_t num_entries_;
+  static const size_t max_entries_ = N;
```

Comment:
There shouldn't be a need for this variable. Rename the template argument if you'd like a longer name for it.

And even if you were computing some value from `N` and needed a variable, it should be `constexpr` (a compile-time constant) rather than `const` (a constant value in memory).